### PR TITLE
DT-813: Switch to webflux and oauth integrated into spring security

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -18,14 +18,12 @@ dependencies {
   runtimeOnly("org.flywaydb:flyway-core:6.4.1")
   runtimeOnly("org.postgresql:postgresql:42.2.12")
 
-  implementation("org.hibernate:hibernate-core:5.4.15.Final")
```
Comment (petergphillips) [inline build.gradle.kts:21]: included in data-jpa anyway

File: build.gradle.kts [hunk 2]
```diff
@@ -50,17 +47,20 @@ dependencies {
...
-  testImplementation("org.springframework.boot:spring-boot-starter-webflux")
   testImplementation("com.tngtech.java:junit-dataprovider:1.13.1")
   testImplementation("net.javacrumbs.json-unit:json-unit-assertj:2.17.0")
   testImplementation("io.github.http-builder-ng:http-builder-ng-apache:1.0.4")
   testImplementation("com.github.tomakehurst:wiremock-standalone:2.26.3")
   testImplementation("org.testcontainers:localstack:1.13.0")
   testImplementation("org.awaitility:awaitility-kotlin:4.0.2")
+  testImplementation("io.jsonwebtoken:jjwt:0.9.1")
+
+  testImplementation("org.springframework.security.oauth:spring-security-oauth2:2.4.1.RELEASE")
```
Comment (petergphillips) [inline build.gradle.kts:63]: only converted production code at present, hence dependencies moved to test scope

File: helm_deploy/offender-case-notes/templates/_envs.tpl [hunk 3]
```diff
@@ -25,6 +25,9 @@ env:
   - name: TOKENVERIFICATION_API_BASE_URL
     value: "{{ .Values.env.TOKENVERIFICATION_API_BASE_URL }}"
 
+  - name: SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_JWK_SET_URI
```
Comment (petergphillips) [inline helm_deploy/offender-case-notes/templates/_envs.tpl:28]: read key from resource server instead

File: src/main/java/uk/gov/justice/hmpps/casenotes/config/ResourceServerConfiguration.java [hunk 4]
```diff
@@ -166,29 +120,18 @@ private Contact contactInfo() {
     }
 
     private ApiInfo apiInfo() {
-        final var vendorExtension = new StringVendorExtension("", "");
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/config/ResourceServerConfiguration.java:169]: fix incorrect generated swagger

File: src/main/java/uk/gov/justice/hmpps/casenotes/config/SecurityUserContext.java [hunk 5]
```diff
@@ -38,32 +34,24 @@ public UserIdUser getCurrentUser() {
 
     private Optional<UserIdUser> getOptionalCurrentUser() {
         final var authentication = getAuthentication();
-        if (authentication == null || authentication.getPrincipal() == null) return Optional.empty();
+        if (!(authentication instanceof AuthAwareAuthenticationToken)) return Optional.empty();
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/config/SecurityUserContext.java:37]: will always be one of our tokens

File: src/main/java/uk/gov/justice/hmpps/casenotes/config/WebClientConfiguration.kt [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  fun tokenVerificationApiHealthWebClient(builder: Builder): WebClient = createHealthClient(builder, prop.tokenVerificationApiBaseUrl)
+
+  private fun createForwardAuthWebClient(builder: Builder, url: @URL String) =
+      builder.baseUrl(url)
+          .filter(addAuthHeaderFilterFunction())
+          .build()
+
+  private fun createHealthClient(builder: Builder, url: @URL String): WebClient {
+    val timeout = prop.healthTimeout
+    val tcpClient = TcpClient.create()
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/config/WebClientConfiguration.kt:54]: bit naff, but apparently this is what you have to do to set a timeout in the webclient (as opposed to setting per request)

File: src/main/resources/application.yml [hunk 7]
```diff
@@ -35,21 +35,32 @@ spring:
...
+  codec:
+    max-in-memory-size: 10MB
+
+
 server:
   port: 8080
   servlet:
     context-path: /
-  use-forward-headers: true
+  forward-headers-strategy: native
```
Comment (petergphillips) [inline src/main/resources/application.yml:59]: new syntax for forward headers - intellij was warning of old syntax

File: src/main/resources/application.yml [hunk 8]
```diff
@@ -35,21 +35,32 @@ spring:
...
   tomcat:
     remote-ip-header: x-forwarded-for
     protocol_header: x-forwarded-proto
 
-logging:
```
Comment (petergphillips) [inline src/main/resources/application.yml:47]: not needed as in logback-spring anyway

File: src/test/java/uk/gov/justice/hmpps/casenotes/config/ClientTrackingTelemetryModuleTest.java [hunk 9]
```diff
@@ -82,22 +82,6 @@ public void shouldAddOnlyClientIdIfUsernameNullToInsightTelemetry() {
 
     }
 
-    @Test
-    public void shouldNotAddClientIdAndUserNameToInsightTelemetryAsTokenExpired() {
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/casenotes/config/ClientTrackingTelemetryModuleTest.java:86]: for telemetry we don't validate the jwt anymore so don't get this exception

File: src/test/java/uk/gov/justice/hmpps/casenotes/health/HealthCheckIntegrationTest.kt [hunk 10]
```diff
@@ -65,9 +65,9 @@ class HealthCheckIntegrationTest : IntegrationTest() {
 
     val response = restTemplate.getForEntity("/health", String::class.java)
 
-    assertThatJson(response.body!!).node("components.OAuthApiHealth.details.error").isEqualTo("org.springframework.web.client.HttpClientErrorException\$NotFound: 404 Not Found: [some error]")
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/casenotes/health/HealthCheckIntegrationTest.kt:68]: different errors for webclient

File: src/test/java/uk/gov/justice/hmpps/casenotes/repository/OffenderCaseNoteRepositoryTest.java [hunk 11]
```diff
@@ -43,6 +46,9 @@
 
     @BeforeEach
     public void setUp() {
+        final var jwt = Jwt.withTokenValue("some").subject("anonymous").header("head", "something").build();
+        SecurityContextHolder.getContext().setAuthentication(
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/casenotes/repository/OffenderCaseNoteRepositoryTest.java:50]: need to actually set with the correct token so user gets injected by the auditor stuff

File: src/test/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteResourceTest.java [hunk 12]
```diff
@@ -152,6 +151,22 @@ public void testCanRetrieveCaseNotesForOffenderNormal() {
         assertThatJsonFileAndStatus(response, 200, "A1234AA-normal-casenote.json");
     }
 
+    @Test
+    public void testRetrieveCaseNotesWillReturn404IfOffenderNotFound() {
+        oauthMockServer.subGetUserDetails("API_TEST_USER");
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteResourceTest.java:156]: missing test for not found that tests controller advice changes

File: src/main/java/uk/gov/justice/hmpps/casenotes/controllers/ControllerAdvice.java [hunk 13]
```diff
@@ -22,16 +22,16 @@
 @Slf4j
 public class ControllerAdvice {
 
-    @ExceptionHandler(RestClientResponseException.class)
-    public ResponseEntity<byte[]> handleRestClientResponseException(final RestClientResponseException e) {
+    @ExceptionHandler(WebClientResponseException.class)
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/controllers/ControllerAdvice.java:25]: cope with new webclient exceptions instead - not think not found should be `log.error`, but replicated existing functionality

File: src/main/java/uk/gov/justice/hmpps/casenotes/health/HealthCheck.kt [hunk 14]
```diff
@@ -4,30 +4,31 @@ import org.springframework.beans.factory.annotation.Qualifier
...
-  }
+abstract class HealthCheck(private val webClient: WebClient) : HealthIndicator {
+  override fun health(): Health? =
+      webClient.get()
+          .uri("/health/ping")
+          .retrieve()
+          .toEntity(String::class.java)
+          .flatMap { Mono.just(Health.up().withDetail("HttpStatus", it?.statusCode).build()) }
+          .onErrorResume(WebClientResponseException::class.java) { Mono.just(Health.down(it).withDetail("body", it.responseBodyAsString).withDetail("HttpStatus", it.statusCode).build()) }
+          .onErrorResume(Exception::class.java) { Mono.just(Health.down(it).build()) }
```
Comment (brightonsbox) [inline src/main/java/uk/gov/justice/hmpps/casenotes/health/HealthCheck.kt:19]: 👍

**Pull request merged**